### DISCUSSION TITLE
record types .ToString() doesn't detect circular references

### DISCUSSION THREAD

**Discussion by efrainsteinbach** [dotnet/csharplang#4390] [General]
### Description

I ran into StackoverflowExceptions caused by infinite loops of ToString() calls along circular references of record types. This is surprising, because to my understanding, record types should be immutable - so in theory circular references should not be possible. (I observed this with records used as EF entities, and the navigation properties often create circular references). I wondered how this is possible on one hand (this may be my bad), and on the other, I would suggest **putting in place a circular reference detection in the generated ToString() code**. In particular, because the VS debugger apparently evaluates all local variables .ToString() when hitting a breakpoint.

This can easily be reproduced:

    class Program
    {
        static void Main(string[] args)
        {
            var r1 = new MyRecord() { Name = "Record 1!", MyRecordRef = null };
            var r2 = new MyRecord() { Name = "Record 2!", MyRecordRef = r1 };
            r1.MyRecordRef = r2; // Shouldn't the compiler prevent this? Sorry, I'm puzzled.

            Console.WriteLine(r1.ToString()); // This dies
            Console.ReadKey();
        }

        public record MyRecord
        {
            public string Name { get; set; }
            public MyRecord MyRecordRef { get; set; }
        }
    }

### Configuration
.net 5.0.102 / Windows 10 / x64

**Comment by CyrusNajmabadi** [comment]
THis is by design.  There is no cycle detection in records in any operation (hashing, equality, stringifying).  If you want to support cyclic records, you need to provide the code to determine what happens here.

**Comment by 333fred** [comment]
> This is surprising, because to my understanding, record types should be immutable

An additional point: record types are immutable _by default_. It is certainly possible, as you can see, to make mutable record types, and when you do so you need to be careful of things like this. This is again by design.

**Comment by pinkfloydx33** [comment]
In response to your "shouldn't the compiler prevent this" comment, `MyRecordRef` has a `set` accessor and not an `init` which is why that line works
